// dnd-core/dist/reducers/index.js
import { get } from '../utils/js_utils.js';
import { reduce as reduce$1 } from './dirtyHandlerIds.js';
import { reduce as reduce$2 } from './dragOffset.js';
import { reduce as reduce$4 } from './dragOperation.js';
import { reduce as reduce$3 } from './refCount.js';
import { reduce as reduce$5 } from './stateId.js';

function _defineProperty(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
function _objectSpread(target) {
    for(var i = 1; i < arguments.length; i++){
        var source = arguments[i] != null ? arguments[i] : {};
        var ownKeys = Object.keys(source);
        if (typeof Object.getOwnPropertySymbols === 'function') {
            ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function(sym) {
                return Object.getOwnPropertyDescriptor(source, sym).enumerable;
            }));
        }
        ownKeys.forEach(function(key) {
            _defineProperty(target, key, source[key]);
        });
    }
    return target;
}
function reduce(state = {}, action) {
    return {
        dirtyHandlerIds: reduce$1(state.dirtyHandlerIds, {
            type: action.type,
            payload: _objectSpread({}, action.payload, {
                prevTargetIds: get(state, 'dragOperation.targetIds', [])
            })
        }),
        dragOffset: reduce$2(state.dragOffset, action),
        refCount: reduce$3(state.refCount, action),
        dragOperation: reduce$4(state.dragOperation, action),
        stateId: reduce$5(state.stateId)
    };
}

export { reduce };
